daemon/smartcard.c

```diff
@@ -177,6 +177,33 @@ smartcard_send_recv_block(const DaemonToToken *out)
 	return msg;
 }
 
+bool
+smartcard_cert_has_valid_format(unsigned char *cert_buf, size_t cert_buf_len)
+{
+	const char *begin_cert_str = "-----BEGIN CERTIFICATE-----\n";
+	const char *end_cert_str = "-----END CERTIFICATE-----\n";
+	size_t begin_cert_str_len = strlen(begin_cert_str);
+	size_t end_cert_str_len = strlen(end_cert_str);
+
+	if (cert_buf == NULL || cert_buf_len == 0) {
+		ERROR("Certificate not provided.");
+		return false;
+	}
+	if (cert_buf_len < end_cert_str_len + begin_cert_str_len) {
+		ERROR("Invalid certificate length %zu.", cert_buf_len);
+		return false;
+	}
+	if (!mem_starts_with(cert_buf, cert_buf_len, begin_cert_str, begin_cert_str_len)) {
```

Comment:
The mem_starts_with / ends_with functions seem to be quite specific to this function.
Therefore, i would suggest to directly place the comparing code here

Reply:
Check out the latest commit, I reworked `smartcard_cert_has_valid_format` to contain the memcmp.